src/vmm/src/vstate/vm.rs

```diff
@@ -355,16 +368,22 @@ impl Vm {
         guest_mem
             .iter()
             .enumerate()
-            .try_for_each(|(index, region)| {
+            .try_for_each(|(_index, region)| {
                 let memory_region = kvm_userspace_memory_region {
-                    slot: index as u32,
+                    slot: self.next_memory_slot as u32,
                     guest_phys_addr: region.start_addr().raw_value() as u64,
                     memory_size: region.len() as u64,
                     // It's safe to unwrap because the guest address is valid.
                     userspace_addr: guest_mem.get_host_address(region.start_addr()).unwrap() as u64,
                     flags,
                 };
 
+                // This is needed because virtio memory devices introduce a new `GuestMemoryMmap`
+                // for each memory device. And each one of those starts numbering the regions
+                // from 0, but they each must be registered with KVM using unique
+                // slot numbers.
+                self.next_memory_slot += 1;
```

Comment:
This needs to be incremented elsewhere because of the reason in the comment above.